crates/driver/src/infra/solver/protocol_fees.rs

```diff
@@ -0,0 +1,270 @@
+                            .unwrap_or_default();
+                    }
+                    Side::Sell => {
+                        // increase buy amount by factor
+                        available.buy.amount = available
+                            .buy
+                            .amount
+                            .apply_factor(1.0 / (1.0 - factor))
+                            .unwrap_or_default();
+                    }
+                }
+            }
+        }
+        Order {
+            uid: order.uid.into(),
+            sell_token: available.sell.token.into(),
+            buy_token: available.buy.token.into(),
+            sell_amount: available.sell.amount.into(),
+            buy_amount: available.buy.amount.into(),
+            fee_amount: available.user_fee.into(),
+            kind: order.side.into(),
+            partially_fillable: order.is_partial(),
+            class: order.kind.into(),
+            fee_policies: self.protocol_fee_to_solvers(weth.into(), order.protocol_fees.as_ref()),
+        }
+    }
+
+    pub fn apply_to_solution(&self, solution: Solution) -> Result<Solution, Error> {
+        let mut trades = Vec::with_capacity(solution.trades.len());
+        for trade in solution.trades {
+            match &trade {
+                Trade::Fulfillment(fulfillment) => match fulfillment.order().kind {
+                    order::Kind::Market | order::Kind::Limit { .. } => {
+                        // Apply protocol fees only if the drivers is set to handler the fees
+                        if self.driver_must_apply_protocol_fee(
+                            fulfillment.order.signature.signer.into(),
+                        ) {
+                            let prices = ClearingPrices {
+                                sell: solution.prices
+                                    [&fulfillment.order().sell.token.wrap(solution.weth)],
+                                buy: solution.prices
+                                    [&fulfillment.order().buy.token.wrap(solution.weth)],
+                            };
+                            let fulfillment = fulfillment
+                                .with_protocol_fee(prices)
+                                .map_err(Error::FailedToApplyProtocolFees)?;
+                            trades.push(Trade::Fulfillment(fulfillment))
+                        } else {
+                            trades.push(trade);
+                        }
+                    }
+                    order::Kind::Liquidity => {
+                        trades.push(trade);
+                    }
+                },
+                Trade::Jit(_) => trades.push(trade),
+            }
+        }
+        Ok(Solution { trades, ..solution })
+    }
+
+    /// Internal function in `ProtocolFees` in order to determine whether the
+    /// driver has to apple the protocol fee to an order/solution
+    fn driver_must_apply_protocol_fee(&self, address: H160) -> bool {
```

Comment:
we have the internal functions `driver_must_apply_protocol_fee()` and `protocol_fee_to_solvers()` because those cases are separated, we may want to apply the protocol fees and not send them to the solver, but also we may want not to apply the protocol fees and not to send them to the solvers.